src/net.cpp

```diff
@@ -2571,7 +2590,18 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect, Spa
             }
         }
 
-        CSemaphoreGrant grant(*semOutbound);
+        CSemaphoreGrant grant{*semOutbound, /*fTry=*/true};
+        if (open_private_broadcast_to.has_value() && grant) {
+            // Previous connection was private broadcast, thus yield to other connection types
+            // if we are ready to do so (the outbound semaphore gave green light).
+            open_private_broadcast_to.reset();
+        } else if (!(open_private_broadcast_to = m_private_broadcast.ShouldOpen(num_private_broadcast_opened))) {
```

Comment:
This code in `CConnman::ThreadOpenConnections()` is not checking whether private broadcast is enabled because it is just a servant, its logic is: "if there is a request to open private broadcast connections, then I will open them". The higher level code is making the decision (currently in RPC only, but eventually in the wallet as well) - "if private broadcast is enabled, then open such connections because I am now broadcasting a transaction (and don't put it in the mempool)". This is why there is no check in `net.cpp` whether `-privatebroadcast` is enabled.

I also find this snippet non-trivial to understand:

```cpp
CSemaphoreGrant grant{*semOutbound, /*fTry=*/true};
if (open_private_broadcast_to.has_value() && grant) {
    // Previous connection was private broadcast, thus yield to other connection types
    // if we are ready to do so (the outbound semaphore gave green light).
    open_private_broadcast_to.reset();
} else if (!(open_private_broadcast_to = m_private_broadcast.ShouldOpen(num_private_broadcast_opened))) {
    // No need for private broadcast connections, go with other ones. Acquire the
    // grant if not already acquired. 
    grant.Acquire();
}
// else private broadcast is needed to network *open_private_broadcast_to, try to open one.
```

Let me summarize why the logic is like that:
* since private broadcast connections are needed now and are short lived, they override all others
* however if there is a lot of demand for private broadcast connections we want to avoid starving other connection types, so we don't open two consecutive private broadcast connections and instead yield to others
* but this yield to others may end up blocking on the grant for a long time because we are not ready to open other connection types whereas private broadcast connections are needed now.

I am open to suggestions on how to implement that in a simpler way or how to simplify the logic itself (without sacrificing the functionality).

Reply:
I didn't mean to suggest a change to the logic, just to make it more conditional on `-privatebroadcast`, so that in principle we wouldn't even need to initiate a `m_private_broadcast` object unless the user is running in `-privatebroadcast` mode. But I don't have a concrete suggestion right now (I might suggest on at a later point but need to look deeper into the grant logic for that) so it's fine to resolve for now.